recipes/md/trajectoryanalysis.py

```diff
@@ -30,19 +36,109 @@ def _get_max_dt_frames(self, max_correlation_time_fs):
         if max_correlation_time_fs is None:
             return None
 
-        historylength = self.previous_job.results.readrkf("History", "nEntries")
-        max_dt_frames = int(max_correlation_time_fs / self.previous_job.results.get_time_step())
+        # Adjust for multiple previous_jobs
+        prevjobs = self.previous_job
+        if not isinstance(prevjobs, list):
+            prevjobs = [prevjobs]
+
+        # Read history and time step
+        # FIXME: Does not account for the presence of Range of StepSize (which the user may have provided)
+        historylength = 0
+        for prevjob in prevjobs:
+            historylength += prevjob.results.readrkf("History", "nEntries")
+        timestep = prevjobs[0].results.get_time_step()
+        max_dt_frames = int(max_correlation_time_fs / timestep)
         max_dt_frames = min(max_dt_frames, historylength // 2)
         return max_dt_frames
 
-    def _parent_prerun(self, section):
+    def get_input(self):
+        """
+        Generate the input file
+        """
+        self._settings_to_list()
+        if self.atom_indices and self._parent_write_atoms:
+            section = getattr(self.settings.input, self._task)
+            for entry in section:
+                found = self._has_settings_entry(entry, "Atoms")
+                if found:
+                    found = self._has_settings_entry(entry.Atoms, "Atom")
+                if not found:
+                    # In case of PISA object
+                    if not isinstance(entry, Settings):
+                        for i, iat in enumerate(self.atom_indices):
+                            entry.Atoms.Atom[i] = self.atom_indices[i]
+                    else:
+                        entry.Atoms.Atom = self.atom_indices
+        return super().get_input()
+
+    def _settings_to_list(self):
+        """
+        Convert the main settings object for this Task to a list of settings objects
+        """
+        if isinstance(self.settings.input, Settings):
```

Comment:
To Matti's point, for this class it might just be easier to use:

```
        if _has_scm_pisa and isinstance(self.settings.input, DriverBlock):
            self.settings.input = self.settings.input.to_settings()
```

or similar in the c'tor, to remove all the if/else.

Trade off would be that the user would not get the benefits of PISA when accessing the job settings...

Reply:
Yes I think PISA should be improved. Could you please raise JIRA(s) with the examples of the bugs/methods to add?

In particular I agree the `contains` is a big omission at the moment and should behave as the `Settings` contains.

I also think that the PISA/Settings objects should be more generally interchangeable to facilitate this kind of refactor. In an ideal world you shouldn't have to make this change at all... (or at least it should be very minimal)